src/Microsoft.AspNetCore.SignalR.Client.TS/Formatters.ts

```diff
@@ -0,0 +1,146 @@
+        }
+
+        // Parse the header
+        var type = knownTypes[header];
+        if (type === undefined) {
+            return ["Unknown type value: '" + header + "'", null];
+        }
+
+        if (type == MessageType.Binary) {
+            // We need to decode and put in an ArrayBuffer. Throw for now
+            // This will require our own Base64-decoder because the browser
+            // built-in one only decodes to strings and throws if invalid UTF-8
+            // characters are found.
+            throw "TODO: Support for binary messages";
+        }
+
+        // Create the message
+        return [null, new Message(type, payload)];
+    }
+}
+
+export namespace Text {
+    const InvalidPayloadError = new Error("Invalid text message payload");
+    const LengthRegex = /^[0-9]+$/;
+
+    function hasSpace(input: string, offset: number, length: number): boolean {
+        let requiredLength = offset + length;
+        return input.length >= requiredLength;
+    }
+
+    function parseMessage(input: string, position: number, log?: (message: string) => void): [string, number, Message] {
+        var offset = position;
+
+        // Read the length
+        var [lenStr, offset] = splitAt(input, ":", offset);
+
+        // parseInt is too leniant, we need a strict check to see if the string is an int
+
+        if (!LengthRegex.test(lenStr)) {
+            return [`Invalid length: '${lenStr}'`, position, null];
+        }
+        let length = Number.parseInt(lenStr);
+
+        // Required space is: 3 (type flag, ":", ";") + length (payload len)
+        if (!hasSpace(input, offset, 3 + length)) {
+            return ["Message is incomplete", position, null];
+        }
+
+        // Read the type
+        var [typeStr, offset] = splitAt(input, ":", offset);
+
+        // Parse the type
+        var type = knownTypes[typeStr];
+        if (type === undefined) {
+            return ["Unknown type value: '" + typeStr + "'", position, null];
+        }
+
+        // Read the payload
+        var payload = input.substr(offset, length);
+        offset += length;
+
+        // Verify the final trailing character
+        if (input[offset] != ';') {
+            return ["Message missing trailer character", position, null]
+        }
+        offset += 1;
+
+        if (type == MessageType.Binary) {
+            // We need to decode and put in an ArrayBuffer. Throw for now
+            // This will require our own Base64-decoder because the browser
+            // built-in one only decodes to strings and throws if invalid UTF-8
+            // characters are found.
+            throw "TODO: Support for binary messages";
+        }
+
+        return [null, offset, new Message(type, payload)];
+    }
+
+    export function parseMessages(input: string): [string, Message[]] {
+        if (input.length == 0) {
+            return [null, []];
+        }
+
+        let messages = [];
+        var offset = 0;
```

Comment:
So I can do it, but it means messing with the return values from the `splitAt` helper. The problem is this pair of lines:

```typescript
let offset = 0;
let [someStr, offset] = splitAt(input, ":", offset);
```

We're trying to introduce a new variable to hold the string before the split and reassign the offset that's tracking our current position in the string (which we are using to avoid allocating too many strings).

Now, we can do this:

```typescript
let offset = 0;

let someStr;
[someStr, offset] = splitAt(input, ":", offset);
```

That looks slightly better than many of the alternatives. Or we could return just the offset, instead of using a tuple, but then we basically have to inline the entire helper and go back to using `indexOf` directly.

Finally, we could introduce an interface rather than using a tuple and do something like:

```typescript
export interface SplitResult {
    prefix: string,
    offset: number
};

let offset = 0;
let result = splitAt(input, 0, offset);
offset = result.offset;
```

Using `var` removes any of these concerns since `var` allows you to redeclare the same variable (which just overwrites it)